modules/core/shared/src/main/scala/io/circe/Printer.scala

```diff
@@ -242,12 +255,22 @@ final case class Printer(
     }
   }
 
+  @transient
+  private[this] final val stringWriter: ThreadLocal[StringBuilder] = new ThreadLocal[StringBuilder] {
+    override final def initialValue: StringBuilder = new StringBuilder()
+  }
+
   /**
    * Returns a string representation of a pretty-printed JSON value.
    */
   final def pretty(json: Json): String = {
-    val writer = new StringBuilder()
-    val folder = new PrintingFolder(writer)
+    val writer = if (reuseWriters) {
```

Comment:
This is nice! I can open a follow-up PR to add the same machinery for `AppendableByteBuffer`.

It's going to me a bit more complicated given it also involves a `Charset`, but I think we can start with caching only UTF-8 writers (like 99% of the use cases).